services/galley/src/Galley/API/MLS/Welcome.hs

```diff
@@ -78,33 +87,34 @@ welcomeRecipients ::
      ]
     r =>
   Welcome ->
-  Sem r [Qualified (UserId, ClientId)]
+  Sem r [(Qualified (UserId, ClientId), Qualified ConvId)]
 welcomeRecipients =
-  traverse
-    ( fmap cidQualifiedClient
-        . derefKeyPackage
-        . gsNewMember
-    )
-    . welSecrets
+  traverse (either throw pure <=< resolveMember . gsNewMember) . welSecrets
+
+resolveMember :: Member BrigAccess r => KeyPackageRef -> Sem r (Either (Tagged 'MLSKeyPackageRefNotFound ()) (Qualified (UserId, ClientId), Qualified ConvId))
+resolveMember kpref = runError $ do
+  clientIdentity <- derefKeyPackage kpref
+  -- TODO Generate new + Put conversaion id?
+  let defConv = Qualified (selfConv (ciUser clientIdentity)) (ciDomain clientIdentity)
+  convId <- fromMaybe defConv <$> getConvIdByKeyPackageRef kpref
+  pure (cidQualifiedClient clientIdentity, convId)
```

Comment:
This seems odd (and has a TODO note). For a given key package ref, the `mls_key_package_refs` table in Brig has the corresponding qualified conversation ID, doesn't it?

Reply:
Not necessarily.  The conversation id fields are optional for that table.  After claiming a key package ref, a conversation id can be added to the claim, but it's not required in order to claim.  That's why the relevant brig internal servant endpoint returns a Maybe.